### PR TITLE
Refactor token updates

### LINKED ISSUE #321: Refactor the hundreds of updateToken calls
In any scenario we make an HTTP call to the API server that requires authentication we pass a token. This token is short lived though, expiring in just one hour. Therefore we need to frequently refresh it prior to making any API calls.

Currently, every component that makes some type of authenticated API request has a copy pasted `updateToken` method within it that looks like this: 

```js
      async updateToken () {
        const token = await auth.currentUser.getIdToken( true );
        this.$axios.setToken( token, 'Bearer' );
      },
```

I've copied pasted this exact function I don't even know how many times, but it's a lot. Enough for even me to think that it's bad.

This is concerning.

This should really be factored out. There's a lot of ways this could be done, but one method is that we could create a custom axios function that triggers this automatically whenever used. e.g. `this.$axios.authenticatedGet` / `this.$axios.authenticatedPost` etc. 

aka this basically:

```js
  const getIdToken = () => {
    if (process.server) {
      if (!req.headers.cookie) { return null }
      return import('cookie').then((cookie) => {
        const parsedCookies = cookie.parse(req.headers.cookie);
        if (!parsedCookies) { return; }
        return parsedCookies[COOKIE_NAME];
      });
    }
    if (process.client) {
      return new Promise((resolve, reject) => {
        const unsubscribe = app.$fireAuth.onAuthStateChanged((user) => {
          unsubscribe();
          if (user) {
            user.getIdToken().then((idToken) => {
              resolve(idToken);
            }, () => {
              resolve(null);
            });
          } else {
            resolve(null);
          }
        });
      });
    }
    return null
  };

  $axios.onRequest(async (config) => {
    if (process.env.NODE_ENV !== 'production') {
      config.withCredentials = true;
    }

    const idToken = await getIdToken();

    if (config.headers != null && config.headers['X-Requested-With'] == null) {
      config.headers = {
        'X-Requested-With': 'XMLHttpRequest',
        Authorization: 'Bearer ' + idToken,
        ...config.headers
      };
    }
    return config;
  });
```

see: https://levelup.gitconnected.com/using-firebase-authentication-in-a-nuxt-server-side-rendered-application-c2a624a9e999 for more info, cuz that's where I stole it from.

But if that's too complicated, it would also make sense to make it a action in the store or an injected function, or it's own plugin, or just a single helper function in `assets/js` ultimately some magic on the axios object is probably "the way" to do it, and everything other than that is equally 2nd place.

Could probably take a lesson or two from how authors of auth plugins organized and structured their auth management though. Also at this point it doesn't have to set a cookie or worry about that because I ripped that off the server to try and improve server response and load time. Later on should go back and look at detecting auth state (server side, we already do it client side) in middleware for guarded routes only so that loading `https://bitwave.tv/profile` directly doesn't require the round trip to the `/login` page and back on initial load.

Reference: lots of calls to updateToken 😥
image

This isn't really a *difficult* problem to solve, but if you wanted to experiment with the axios stuff, it could be time consuming to figure out how that works, what does and doesn't, and ensuring it isn't (much) slower.

### PR DESCRIPTION
**Pull request by diingus**
Gone are the days of copy-pasted `getTokenId()` or `getFreshTokenId()` definitions and calls.

Surprisingly, all token usage was for request done through Axios. Axios requests are now intercepted and the Authorization header is being injected. The auth token is also automatically refreshed on client-side.

Server-side token parsing-from-cookies is also possible, but the commit-to-store part of it is commented out. Should only commit when on auth-guarded pages, or pages that make little sense to cache.

Closes #321

### CODE REVIEW

**Comment by diingus** [comment]
I remember thinking of an issue with this branch last night before falling asleep. It could have also been a dream :thinking: 

Give me a few to try remember it

**Comment by diingus** [comment]
Likely a dream, since I can't seem to remember it, and I can't find it in code.

File: plugins/firebase.js [hunk 1]
```diff
@@ -134,6 +135,62 @@ export default async ( { app, store }, inject ) => {
...
+      }
+      if (process.client) {
+        // Promise soup. It likely works.
+        return new Promise((resolve, reject) => {
+          const unsubscribe = listenToAuthState((user) => {
+            unsubscribe();
+            if (user) {
+              // If within 5 mins before expiration, force a token refresh
+              // Otherwise, return what you already got (idToken)
+              user.getIdToken().then( async (idToken) => {
```
Comment (DispatchCommit) [inline plugins/firebase.js:160]: just noting that you can replace `.then()`'s with `await`'s but I'll handle that post merge.

File: plugins/firebase.js [hunk 2]
```diff
@@ -134,6 +135,62 @@ export default async ( { app, store }, inject ) => {
...
+    // Intercepts all axios requests and injects Bearer token
+    //  into the Authorization header. Equivalent to setToken(), except
+    //  it gets called implicitly on each request.
+    $axios.onRequest(async (config) => {
+      const idToken = await getIdToken();
+
+      if (config.headers != null && config.headers['X-Requested-With'] == null) {
+        config.headers = {
+          'X-Requested-With': 'XMLHttpRequest',
+          Authorization: 'Bearer ' + idToken,
```
Comment (DispatchCommit) [inline plugins/firebase.js:187]: Should the header be added if the `idToken` is `null`?

**Review by DispatchCommit — CHANGES_REQUESTED** [review]

**Review by DispatchCommit — APPROVED** [review]

**Pull request merged**